File: lib/go/internal/SchemaIntrospection.go
//|
//|  Copyright The Telepact Authors
//|
//|  Licensed under the Apache License, Version 2.0 (the "License");
//|  you may not use this file except in compliance with the License.
//|  You may obtain a copy of the License at
//|
//|  https://www.apache.org/licenses/LICENSE-2.0
//|
//|  Unless required by applicable law or agreed to in writing, software
//|  distributed under the License is distributed on an "AS IS" BASIS,
//|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//|  See the License for the specific language governing permissions and
//|  limitations under the License.
//|

package internal

import (
	"regexp"
	"strings"
)

func GetIndexEntries(schema SchemaAccessor, includeInternal bool) []any {
	definitions := schema.OriginalDefinitions()
	if includeInternal {
		definitions = schema.FullDefinitions()
	}

	entries := make([]any, 0)
	for _, definition := range definitions {
		definitionMap := toStringAnyMap(definition)
		schemaKey := getSchemaKey(definitionMap)
		if !strings.HasPrefix(schemaKey, "fn.") || strings.HasSuffix(schemaKey, ".->") {
			continue
		}
		if !includeInternal && strings.HasSuffix(schemaKey, "_") {
			continue
		}

		entry := map[string]any{"name": schemaKey}
		if comment, ok := definitionMap["///"]; ok {
			entry["comment!"] = comment
		}
		entries = append(entries, entry)
	}

	return entries
}

func GetDefinitionClosure(schema SchemaAccessor, name string, includeInternal bool) []any {
	rootDefinition := getRootDefinition(schema, name, includeInternal)
	if rootDefinition == nil {
		return []any{}
	}

	definitionsByName := map[string]map[string]any{}
	for _, definition := range schema.FullDefinitions() {
		definitionMap := toStringAnyMap(definition)
		definitionsByName[getSchemaKey(definitionMap)] = definitionMap
	}

	visited := map[string]bool{}
	var visit func(schemaKey string)
	visit = func(schemaKey string) {
		if visited[schemaKey] {
			return
		}

		definition, ok := definitionsByName[schemaKey]
		if !ok {
			return
		}

		visited[schemaKey] = true
		for _, reference := range getDefinitionReferences(definition, definitionsByName) {
			visit(reference)
		}
	}

	visit(getSchemaKey(rootDefinition))

	closure := make([]any, 0)
	for _, definition := range schema.FullDefinitions() {
		definitionMap := toStringAnyMap(definition)
		if visited[getSchemaKey(definitionMap)] {
			closure = append(closure, definition)
		}
	}

	return closure
}

func getRootDefinition(schema SchemaAccessor, name string, includeInternal bool) map[string]any {
	if !includeInternal && strings.HasSuffix(name, "_") {
		return nil
	}

	definitions := schema.OriginalDefinitions()
	if includeInternal {
		definitions = schema.FullDefinitions()
	}

	for _, definition := range definitions {
		definitionMap := toStringAnyMap(definition)
		if getSchemaKey(definitionMap) == name {
			return definitionMap
		}
	}

	return nil
}

func getDefinitionReferences(definition map[string]any, definitionsByName map[string]map[string]any) []string {
	references := map[string]bool{}
	schemaKey := getSchemaKey(definition)

	for key, value := range definition {
		switch key {
		case "///":
			continue
		case "_errors":
			regexString, ok := value.(string)
			if !ok {
				continue
			}
			regex, err := regexp.Compile(regexString)
			if err != nil {
				continue
			}
			for candidate := range definitionsByName {
				if regex.MatchString(candidate) {
					references[candidate] = true
				}
			}
		case schemaKey, "->":
			for _, reference := range getTypeExpressionReferences(value, definitionsByName) {
				references[reference] = true
			}
		}
	}

	result := make([]string, 0, len(references))
	for reference := range references {
		result = append(result, reference)
	}
	return result
}

func getTypeExpressionReferences(typeExpression any, definitionsByName map[string]map[string]any) []string {
	references := map[string]bool{}

	switch value := typeExpression.(type) {
	case string:
		schemaKey := strings.TrimSuffix(value, "?")
		if _, ok := definitionsByName[schemaKey]; ok {
			references[schemaKey] = true
		}
	case []any:
		if isUnionDefinition(value) {
			for _, tagDefinition := range value {
				tagMap := toStringAnyMap(tagDefinition)
				for key, entry := range tagMap {
					if key == "///" {
						continue
					}
					for _, reference := range getTypeExpressionReferences(entry, definitionsByName) {
						references[reference] = true
					}
				}
			}
		} else if len(value) > 0 {
			for _, reference := range getTypeExpressionReferences(value[0], definitionsByName) {
				references[reference] = true
			}
		}
	case map[string]any:
		if len(value) == 1 {
			if objectValue, ok := value["string"]; ok {
				for _, reference := range getTypeExpressionReferences(objectValue, definitionsByName) {
					references[reference] = true
				}
				break
			}
		}
		for _, entry := range value {
			for _, reference := range getTypeExpressionReferences(entry, definitionsByName) {
				references[reference] = true
			}
		}
	}

	result := make([]string, 0, len(references))
	for reference := range references {
		result = append(result, reference)
	}
	return result
}

func isUnionDefinition(typeExpression []any) bool {
	if len(typeExpression) == 0 {
		return false
	}

	for _, entry := range typeExpression {
		tagMap := toStringAnyMap(entry)
		if tagMap == nil {
			return false
		}

		nonCommentKeys := 0
		for key, value := range tagMap {
			if key == "///" {
				continue
			}
			if toStringAnyMap(value) == nil {
				return false
			}
			nonCommentKeys += 1
		}
		if nonCommentKeys != 1 {
			return false
		}
	}

	return true
}

File: lib/ts/src/internal/SchemaIntrospection.ts
//|
//|  Copyright The Telepact Authors
//|
//|  Licensed under the Apache License, Version 2.0 (the "License");
//|  you may not use this file except in compliance with the License.
//|  You may obtain a copy of the License at
//|
//|  https://www.apache.org/licenses/LICENSE-2.0
//|
//|  Unless required by applicable law or agreed to in writing, software
//|  distributed under the License is distributed on an "AS IS" BASIS,
//|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//|  See the License for the specific language governing permissions and
//|  limitations under the License.
//|

import { TelepactSchema } from '../TelepactSchema.js';

export function getIndexEntries(telepactSchema: TelepactSchema, includeInternal: boolean): Record<string, any>[] {
    const definitions = includeInternal ? telepactSchema.full : telepactSchema.original;

    return definitions
        .map((definition) => definition as Record<string, any>)
        .map((definition) => ({ definition, schemaKey: getSchemaKey(definition) }))
        .filter(({ schemaKey }) => schemaKey.startsWith('fn.') && !schemaKey.endsWith('.->'))
        .filter(({ schemaKey }) => includeInternal || !schemaKey.endsWith('_'))
        .map(({ definition, schemaKey }) => {
            const entry: Record<string, any> = { name: schemaKey };
            if (definition['///'] !== undefined) {
                entry['comment!'] = definition['///'];
            }
            return entry;
        });
}

export function getDefinitionClosure(
    telepactSchema: TelepactSchema,
    name: string,
    includeInternal: boolean,
): any[] {
    const definitionsByName = Object.fromEntries(
        telepactSchema.full.map((definition) => {
            const definitionRecord = definition as Record<string, any>;
            return [getSchemaKey(definitionRecord), definitionRecord];
        }),
    ) as Record<string, Record<string, any>>;
    const rootDefinition = getRootDefinition(telepactSchema, name, includeInternal);
    if (!rootDefinition) {
        return [];
    }

    const visited = new Set<string>();
    const visit = (schemaKey: string) => {
        if (visited.has(schemaKey)) {
            return;
        }

        const definition = definitionsByName[schemaKey];
        if (!definition) {
            return;
        }

        visited.add(schemaKey);
        for (const reference of getDefinitionReferences(definition, definitionsByName)) {
            visit(reference);
        }
    };

    visit(getSchemaKey(rootDefinition));

    return telepactSchema.full.filter((definition) =>
        visited.has(getSchemaKey(definition as Record<string, any>)),
    );
}

function getRootDefinition(
    telepactSchema: TelepactSchema,
    name: string,
    includeInternal: boolean,
): Record<string, any> | undefined {
    if (!includeInternal && name.endsWith('_')) {
        return undefined;
    }

    const definitions = includeInternal ? telepactSchema.full : telepactSchema.original;
    return definitions
        .map((definition) => definition as Record<string, any>)
        .find((definition) => getSchemaKey(definition) === name);
}

function getDefinitionReferences(
    definition: Record<string, any>,
    definitionsByName: Record<string, Record<string, any>>,
): Set<string> {
    const references = new Set<string>();
    const schemaKey = getSchemaKey(definition);

    for (const [key, value] of Object.entries(definition)) {
        if (key === '///') {
            continue;
        }

        if (key === '_errors') {
            if (typeof value === 'string') {
                const regex = new RegExp(value);
                for (const candidate of Object.keys(definitionsByName)) {
                    if (regex.test(candidate)) {
                        references.add(candidate);
                    }
                }
            }
            continue;
        }

        if (key === schemaKey || key === '->') {
            for (const reference of getTypeExpressionReferences(value, definitionsByName)) {
                references.add(reference);
            }
        }
    }

    return references;
}

function getTypeExpressionReferences(
    typeExpression: any,
    definitionsByName: Record<string, Record<string, any>>,
): Set<string> {
    const references = new Set<string>();

    if (typeof typeExpression === 'string') {
        const schemaKey = typeExpression.endsWith('?')
            ? typeExpression.slice(0, -1)
            : typeExpression;
        if (definitionsByName[schemaKey]) {
            references.add(schemaKey);
        }
        return references;
    }

    if (Array.isArray(typeExpression)) {
        if (isUnionDefinition(typeExpression)) {
            for (const tagDefinition of typeExpression) {
                for (const [key, value] of Object.entries(tagDefinition as Record<string, any>)) {
                    if (key === '///') {
                        continue;
                    }
                    for (const reference of getTypeExpressionReferences(value, definitionsByName)) {
                        references.add(reference);
                    }
                }
            }
        } else if (typeExpression.length > 0) {
            for (const reference of getTypeExpressionReferences(typeExpression[0], definitionsByName)) {
                references.add(reference);
            }
        }
        return references;
    }

    if (typeof typeExpression === 'object' && typeExpression !== null) {
        const objectDefinition = typeExpression as Record<string, any>;
        if (Object.keys(objectDefinition).length === 1 && objectDefinition.string !== undefined) {
            for (const reference of getTypeExpressionReferences(objectDefinition.string, definitionsByName)) {
                references.add(reference);
            }
            return references;
        }

        for (const value of Object.values(objectDefinition)) {
            for (const reference of getTypeExpressionReferences(value, definitionsByName)) {
                references.add(reference);
            }
        }
    }

    return references;
}

function isUnionDefinition(typeExpression: any[]): boolean {
    if (typeExpression.length === 0) {
        return false;
    }

    return typeExpression.every((entry) => {
        if (typeof entry !== 'object' || entry === null || Array.isArray(entry)) {
            return false;
        }

        const nonCommentKeys = Object.keys(entry).filter((key) => key !== '///');
        return nonCommentKeys.length === 1 && typeof (entry as Record<string, any>)[nonCommentKeys[0]] === 'object';
    });
}

function getSchemaKey(definition: Record<string, any>): string {
    for (const key of Object.keys(definition)) {
        if (key !== '///' && key !== '->' && key !== '_errors') {
            return key;
        }
    }

    throw new Error(`Schema entry has no schema key: ${JSON.stringify(definition)}`);
}
